### DISCUSSION TITLE
SSO using SAML?

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#16857] [Help]
Hi!

Does anyone have experience implementing SSO using SAML? I have a project that requires the use of SAML as it's SSO solution, OAuth isn't possible for a third-party. The plan is have content in a PHP CMS which possibly could act as a proxy.

I'm leaning towards a flow where the CMS handles all communication with the IdP and only having the session there, then just redirecting back to the app with an OAuth token for the user session in the CMS.

Otherwise I'm looking at making app work as a SP with it's own session handling... The plan is to deploy the app with a server for incremental SSG, so storing sessions server-side isn't going to be a problem. I'm more worried about that this session data is required to be able to fetch data from the CMS for a lot of routes which would lead to the Next API getting lots of requests for session data before fetching stuff from the CMS.

Does anyone have any experience/advice for how to proceed?

**Comment by bkingmosaic** [comment]
I need the same thing. While I have integrated with Auth0 and other identity providers WITHOUT Next.js in the past, I'm confused about what the correct approach is to perform authentication in general in Next.js with standard OAuth2 providers. Anyone know of a good tutorial ?

**Reply by sbauch** [reply]
have you checked out [NextAuth.js](https://github.com/nextauthjs/next-auth)? they support a lot of OAuth2 providers (including Auth0) and have great docs.

I maintain [Osso](https://enterprise-oss/osso), an open source service for managing SAML SSO. It's similar to Auth0, but only for SAML (as opposed to identity as a service).

I have a PR open to add Osso as provider to NextAuth.js - https://github.com/nextauthjs/next-auth/pull/928

I'm running a demo here - https://nextjs-demo.ossoapp.com/

When you choose to sign in to the demo Next app, you're sent to Osso's hosted login page on our [demo Osso instance](https://demo.ossoapp.com). 

The demo instance is configured with a tenant for `example.com`, where that fake company uses [Osso's Mock IDP](https://idp.ossoapp.com) as an Identity Provider service. The Mock IDP will take any password for a user so you can see the whole loging flow for SAML SSO in a NextJS app.

I'm less clear on some of OP's concerns but would be happy to talk through as we try to add better support for JAM stack generally.
